caikit/interfaces/vision/data_model/image.py

```diff
@@ -0,0 +1,111 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Data structures for representing images."""
+
+# Third Party
+from PIL import Image as PILImage
+import numpy as np
+
+# First Party
+from py_to_proto.dataclass_to_proto import Annotated, FieldNumber
+import alog
+
+# Local
+from .backends import ImagePilBackend
+from .package import VISION_PACKAGE
+from caikit.core import DataObjectBase, dataobject, error_handler
+
+log = alog.use_channel("DATAM")
+error = error_handler.get(log)
+
+
+@dataobject(package=VISION_PACKAGE)
+class Image(DataObjectBase):
+    """Data model for an image object; this stores the image in the backend as a PIL image, with
+    convenience views to interact with the Image as other formats as needed.
+    """
+
+    image_data: Annotated[bytes, FieldNumber(1)]
+
+    def __init__(self, *args, **kwargs):
+        # For now, we always delegate to the Numpy image backend, because it's the only one.
+        # In the future, we may have more views etc, but for now, images as numpy arrays is
+        # the hub format.
+        #
+        # NOTE: Currently, the args / kwargs here should match up with the proto class, because
+        # .from_proto() requires it.
+        self._backend = ImagePilBackend(*args, **kwargs)
+
+    def as_numpy(self) -> np.ndarray:
+        """Convert the Image data model to a Numpy Array; here, we delegate to the backend.
+        Produced images are of type uint8.
+        Note that the produced object will have shape (Rows, Cols, Channels).
+        Returns:
+            np.ndarray
+                Numpy array(uint8) representation of this data model object.
+        """
+        self._check_initialization()
+        return self._backend.as_numpy()
+
+    def as_pil(self) -> PILImage.Image:
+        """Convert the Image data model to a PIL image. Since we use PIL images as our hub format,
+        this is simply returning a handle to the internally stored PIL image.
+        Returns:
+            PIL.Image
+                PIL image representation of this data model object.
+        """
+        self._check_initialization()
+        return self._backend.as_pil()
+
+    def _check_initialization(self):
+        """Ensure that we have a _backend; throw an attribute error if we don't. This will always
+        be the case if we go through the initializer, but may occur if we call __new__() and use
+        this class incorrectly.
+        """
+        if not isinstance(getattr(self, "_backend", None), ImagePilBackend):
+            error(
+                "<COR14411171E>",
+                AttributeError(
+                    "Unable to create view; missing PIL backend initialization"
+                ),
+            )
+
+    ### Other attributes
+    @property
+    def rows(self) -> int:
```

Comment:
Good question 😄 initializing stuff from the client is the main reason that I didn't want to do this for now. The primary reason being that this should be encoded in the `image_data`, and giving clients leveraging the protobuf classes in their clients the ability to set these really just gives them the ability to redundant information that might also be conflicting with the compressed image bytes.

This could also be a relevant point for when we add geotiff support / support for other formats though, as I'm not very familiar with how those get encoded 🤔

Reply:
Ok, I'd say leave it as is then since it's easier to add fields later than take them away